compiler/noirc_evaluator/src/ssa/opt/inlining.rs

```diff
@@ -234,56 +289,178 @@ fn compute_times_called(ssa: &Ssa) -> HashMap<FunctionId, usize> {
         })
 }
 
-fn should_retain_recursive(
+/// Compute for each function the set of functions that call it, and how many times they do so.
+fn compute_callers(ssa: &Ssa) -> BTreeMap<FunctionId, BTreeMap<FunctionId, usize>> {
+    ssa.functions
+        .iter()
+        .flat_map(|(caller_id, function)| {
+            let called_functions = called_functions_vec(function);
+            called_functions.into_iter().map(|callee_id| (*caller_id, callee_id))
+        })
+        .fold(
+            // Make sure an entry exists even for ones that don't get called.
+            ssa.functions.keys().map(|id| (*id, BTreeMap::new())).collect(),
+            |mut acc, (caller_id, callee_id)| {
+                let callers = acc.entry(callee_id).or_default();
+                *callers.entry(caller_id).or_default() += 1;
+                acc
+            },
+        )
+}
+
+/// Compute for each function the set of functions called by it, and how many times it does so.
+fn compute_callees(ssa: &Ssa) -> BTreeMap<FunctionId, BTreeMap<FunctionId, usize>> {
+    ssa.functions
+        .iter()
+        .flat_map(|(caller_id, function)| {
+            let called_functions = called_functions_vec(function);
+            called_functions.into_iter().map(|callee_id| (*caller_id, callee_id))
+        })
+        .fold(
+            // Make sure an entry exists even for ones that don't call anything.
+            ssa.functions.keys().map(|id| (*id, BTreeMap::new())).collect(),
+            |mut acc, (caller_id, callee_id)| {
+                let callees = acc.entry(caller_id).or_default();
+                *callees.entry(callee_id).or_default() += 1;
+                acc
+            },
+        )
+}
+
+/// Compute something like a topological order of the functions, starting with the ones
+/// that do not call any other functions, going towards the entry points. When cycles
+/// are detected, take the one which are called by the most to break the ties.
+///
+/// This can be used to simplify the most often called functions first.
+///
+/// Returns the functions paired with their transitive weight, which accumulates
+/// the weight of all the functions they call.
+pub(super) fn compute_bottom_up_order(ssa: &Ssa) -> Vec<(FunctionId, usize)> {
+    let mut order = Vec::new();
+    let mut visited = HashSet::new();
+
+    // Number of times a function is called, to break cycles.
+    let mut times_called = compute_times_called(ssa).into_iter().collect::<Vec<_>>();
```

Comment:
This method returns `HashMap`, we collect into a Vec, and then `sort_by_key`. Can we just return `BTreeMap`?

Reply:
The issue with that might be that I actually sort by the the _value_ first, then the key, so that when I call `.pop()` I get the function which was called the the next most amount of time. The `BTreeMap` would be sorted by key.

Reply:
Oh I see, I didn't read the internals of the `sort_by_key` closure close enough.